internal/agent/config/config.go

```
@@ -2,43 +2,77 @@ package config
-	ReportInterval types.DurationInSeconds `yaml:"reportInterval" env:"REPORT_INTERVAL" env-default:"10"`
-	HashKey        string                  `yaml:"hashKey" env:"KEY"`
-	RateLimit      int                     `yaml:"rateLimit" env:"RATE_LIMIT" env-default:"3"`
+	Addr           string                  `json:"address" env:"ADDRESS" env-default:"localhost:8080"`
+	PollInterval   types.DurationInSeconds `json:"poll_interval" env:"POLL_INTERVAL" env-default:"2"`
+	ReportInterval types.DurationInSeconds `json:"report_interval" env:"REPORT_INTERVAL" env-default:"10"`
+	HashKey        string                  `json:"hash_key" env:"KEY"`
+	RateLimit      int                     `json:"rate_limit" env:"RATE_LIMIT" env-default:"3"`
+	CryptoKey      string                  `json:"crypto_key" env:"CRYPTO_KEY"`
+	ConfigPath     string                  `json:"-"`
 }
 
 func Load() (*Config, error) {
 	cfg := &Config{}
 
-	flag.Func("p", "частота опроса метрик из пакета runtime в секундах", func(s string) error { return cfg.PollInterval.SetValue(s) })
-	flag.Func("r", "частота отправки метрик на сервер в секундах", func(s string) error { return cfg.ReportInterval.SetValue(s) })
-	flag.StringVar(&cfg.HashKey, "k", "", "ключ для хеширования")
-	flag.IntVar(&cfg.RateLimit, "l", 3, "лимит одновременной отправки метрик")
-	flag.StringVar(&cfg.Server.Addr, "a", "localhost:8080", "адрес эндпоинта HTTP-сервера")
+	setupFlags(cfg)
 
-	flag.Parse()
+	if cfg.ConfigPath == "" {
+		cfg.ConfigPath = os.Getenv("CONFIG")
+	}
 
-	err := cleanenv.ReadEnv(cfg)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't read environment variables: %w", err)
+	if cfg.ConfigPath != "" {
+		loadJSONConfig(cfg)
 	}
 
+	applyEnvVars(cfg)
+
+	setupFlags(cfg)
+
 	log.Printf("loaded config: %+v", cfg)
 
 	return cfg, nil
 }
+
+func loadJSONConfig(cfg *Config) {
+	err := cleanenv.ReadConfig(cfg.ConfigPath, cfg)
+	if err != nil {
+		log.Printf("warning: couldn't read JSON config from %s: %s", cfg.ConfigPath, err)
```

Существует хороший приннцип "Fail early": если при инициализации программы что-то пошло не так, то лучше не пытаться исправить ошибку, а просто завершить процесс с ненулевым кодом. Так будет сразу видно, что что-то пошло не так.

Если сделать fallback на дефолтные значения, то это может быть неожиданно для клиента: он передавал один параметром, а сервис запустился с другим. Это может привести к неожиданному (а значит непонятному) поведению сервиса.

Исправлять некритические ошибки во время работы сервиса -- хорошо, при инициализации -- скорее плохо